byzcoin/service_test.go

```diff
@@ -1850,6 +1850,70 @@ func TestService_SetConfigRosterReplace(t *testing.T) {
 	}
 }
 
+// Check consistency of the set of valid peers while replacing roster
+func TestService_CheckValidPeers(t *testing.T) {
+	s := newSer(t, 1, testInterval)
+	defer s.local.CloseAll()
+
+	newServers, newRoster, _ := s.local.MakeSRS(cothority.Suite, 4, ByzCoinID)
+
+	// Compute the peerSetID for this skipchain
+	onetCtx := s.service().ServiceProcessor.Context
+	peerSetID := onetCtx.NewPeerSetID(s.genesis.SkipChainID())
+
+	// List of all servers involved in the test
+	allServers := append(s.hosts, newServers...)
+
+	log.Lvl1("Replace with new roster", newRoster.List)
+	goodRoster := onet.NewRoster(s.roster.List)
+	counter := 1
+
+	// Expected valid peers during the test, initialized to the starting roster.
+	expectedValidPeers := []network.ServerIdentityID{}
+	for _, srv := range goodRoster.List {
+		expectedValidPeers = append(expectedValidPeers, srv.ID)
+	}
+
+	for index, si := range newRoster.List {
+		log.Lvl1("Adding", si)
+		goodRoster = onet.NewRoster(append(goodRoster.List, si))
+		expectedValidPeers = append(expectedValidPeers, si.ID)
+
+		ctx, _ := createConfigTxWithCounter(t, testInterval, *goodRoster, defaultMaxBlockSize, s, counter)
+		counter++
+		cl := NewClient(s.genesis.SkipChainID(), *goodRoster)
+		resp, err := cl.AddTransactionAndWait(ctx, 10)
+		transactionOK(t, resp, err)
+
+		// Wait until all servers have included the block
+		time.Sleep(testInterval * 2)
```

Comment:
Can you replace that with `cl.WaitPropagation(-1)`? It will make sure that all nodes have the same last block present. And as the `AddTransactionAndWait` will not return until at least the contacted node has the new block stored, this will wait the correct time, without having to sleep, which might break if the testing-host is busy with other stuff.

Same below.

And, yes, not all tests use yet `WaitPropagation`. But they should ;)

Reply:
Excellent, thank you! I was indeed hoping there was a way to do that without the brittle `Sleep`...